### PR TITLE
additional GET aggregations

### LINKED ISSUE #65: Improvements to GET
GET currently has only two aggregations: "first" and "sum". Some additional useful aggregations would be:

1. last (property value of last widget in collection)
2. min (lowest property value of widgets in collection)
3. max (highest property value of widgets in collection)
4. average (average property value of widgets in collection)
5. median (median property value of widgets in collection)

GET should also allow to make use of the holder parameter instead of making a prior selection, for sake of consistency and also because it seems quite useful to be able get an aggregated property value of widgets in a holder.

### LINKED ISSUE #381: Add `all` aggregation mode to `GET`
When given a collection, `GET` should have the ability to return an equal-sized collection that contains the property of each element of the source collection.

### PR DESCRIPTION
**Pull request by larkob**
This PR adds 'last', 'array', 'average', 'median', 'min' and 'max' to the GET aggregations. The numerical aggregations (min, max, sum, average, median) treat any numerical strings as numbers (and empty as 0).

I've also introduced a new parameter `skipMissing` which defaults to false, but if set to true removes any null or undefined values prior to the aggregation.

Resolves #65 
Resolves #381

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3383/pr-test (or any other room on that server)

**Comment by ArnoldSmith86** [comment]
`array` should work with any values though, right?

**Comment by larkob** [comment]
Yes

File: client/js/widgets/widget.js [hunk 1]
```diff
@@ -586,29 +586,42 @@ export class Widget extends StateManaged {
...
-            variables[a.variable] = 0;
-            for(const widget of collections[a.collection])
-              variables[a.variable] += Number(widget.p(a.property) || 0);
-            break;
-          default:
-            problems.push(`Aggregation ${a.aggregation} is unsupported.`);
-          }
+          let c = collections[a.collection];
+          if (a.skipMissing)
+            c = c.filter(w=>w.p(a.property));
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:593]: This also filters zeros, right? Feels like only `undefined` and `null` would be more useful.

File: client/js/widgets/widget.js [hunk 2]
```diff
@@ -586,29 +586,42 @@ export class Widget extends StateManaged {
...
+          c = JSON.parse(JSON.stringify(c.map(w=>w.p(a.property))));
+          if(c.length) {
+            switch(a.aggregation) {
+            case 'first':
+            case 'last':
+              const index = (a.aggregation == 'last') ? c.length -1 : 0;
+              if(c[index] !== undefined)
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:600]: Did you want to delete this line? The check is also in the next one.

File: client/js/widgets/widget.js [hunk 3]
```diff
@@ -586,29 +586,42 @@ export class Widget extends StateManaged {
...
+                variables[a.variable] = (c[index] !== undefined) ? c[index] : null;
+              break;
+            case 'array':
+              variables[a.variable] = c;
+              break;
+            case 'average':
+              variables[a.variable] = c.map(w=>+w).reduce((a, b) => +a + +b) / c.length;
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:607]: The map is redundant, isn't it? You convert in reduce as well.

File: client/js/widgets/widget.js [hunk 4]
```diff
@@ -586,29 +586,42 @@ export class Widget extends StateManaged {
...
+              break;
+            case 'median':
+              const mid = Math.floor(c.length / 2);
+              const nums = [...c].map(w=>+w).sort((a, b) => a - b);
+              variables[a.variable] = c.length % 2 !== 0 ? nums[mid] : (nums[mid - 1] + nums[mid]) / 2;
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:612]: The `!== 0` should be unnecessary but might help readability.

File: client/js/widgets/widget.js [hunk 5]
```diff
@@ -586,29 +586,42 @@ export class Widget extends StateManaged {
...
+              break;
+            case 'min':
+            case 'max':
+              variables[a.variable] = Math[a.aggregation](...c);
+              break;
+            case 'sum':
+              variables[a.variable] = 0 + c.map(w=>+w).reduce((a, b) => a + b);
```
Comment (ArnoldSmith86) [inline client/js/widgets/widget.js:619]: Why the `0 +` if everything is numeric anyway?

**Review by ArnoldSmith86 — APPROVED** [review]
Apart from the first comment just nitpicking.

Very nice.

I did not test this at all.

File: client/js/widgets/widget.js (hunk 1 above)
Comment (larkob) [reply client/js/widgets/widget.js]: Right. It now allows for 0 but not null or undefined

File: client/js/widgets/widget.js (hunk 3 above)
Comment (larkob) [reply client/js/widgets/widget.js]: Right, I've removed the redundancy. I want to convert with map for better readability and consistency with the other cases

File: client/js/widgets/widget.js (hunk 4 above)
Comment (larkob) [reply client/js/widgets/widget.js:612]: Agree

File: client/js/widgets/widget.js (hunk 5 above)
Comment (larkob) [reply client/js/widgets/widget.js]: Right, it has been removed.

**Pull request merged**